File: Easy/0559-maximum-depth-of-n-ary-tree/0559-maximum-depth-of-n-ary-tree.js
/**
 * // Definition for a _Node.
 * function _Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {_Node|null} root
 * @return {number}
 */
var maxDepth = function(root) {
    if(root === null){
        return 0;
    }
    
    const queue = [root];
    let depth = 0;
    
    while(queue.length > 0){
        const size = queue.length;
        
        for(let i = 0; i < size; i++){
            const cur = queue.shift();
            
            for(const child of cur.children){
                if(child != null){
                    queue.push(child);
                }
            }
        }
        
        depth++;
    }
    
    return depth;
};

File: Easy/0559-maximum-depth-of-n-ary-tree/0559-maximum-depth-of-n-ary-tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public int maxDepth(Node root) {
        if(root == null){
            return 0;
        }
        
        return maxDepth(root, 1);
    }
    
    public int maxDepth(Node root, int depth){
        if(root == null || root.children.isEmpty()){
            return depth;
        }
                
        int max = depth;
        for(Node child : root.children){
            int nextDepth = maxDepth(child, depth + 1);
            max = Math.max(max, nextDepth);
        }
        
        return max;
    }
}